fairlearn/reductions/exponentiated_gradient/exponentiated_gradient.py

```diff
@@ -0,0 +1,267 @@
+"""
+
+from __future__ import print_function
+
+import numpy as np
+import pandas as pd
+import functools
+from fairlearn.reductions import moments
+from fairlearn.reductions import ReductionsEstimator
+from ._constants import _ACCURACY_MUL, _REGRET_CHECK_START_T, _REGRET_CHECK_INCREASE_T, \
+    _SHRINK_REGRET, _SHRINK_ETA, _MIN_T, _RUN_LP_STEP, _PRECISION, _INDENTATION
+from ._lagrangian import _Lagrangian
+from collections import namedtuple
+
+print = functools.partial(print, flush=True)
+
+
+def _mean_pred(X, hs, weights):
+    """Return a weighted average of predictions produced by classifiers in hs"""
+    pred = pd.DataFrame()
+    for t in range(len(hs)):
+        pred[t] = hs[t](X)
+    return pred[weights.index].dot(weights)
+
+
+# TODO: decide whether we want a structured object instead of the namedtuple
+column_names = "best_classifier best_gap classifiers weights last_t best_t n_oracle_calls"
+expgrad_result = namedtuple("ExpgradResult", column_names)
+
+
+class ExponentiatedGradientResult:
+    def __init__(self, best_classifier, best_gap, classifiers, weights, last_t, best_t,
+                 n_oracle_calls):
+        self._best_classifier = best_classifier
+        self._best_gap = best_gap
+        self._classifiers = classifiers
+        self._weights = weights
+        self._last_t = last_t
+        self._best_t = best_t
+        self._n_oracle_calls = n_oracle_calls
+
+    def _predict_with_best_classifier(self, X):
+        return self._best_classifier(X)
+
+    def _asdict(self):
+        return {
+            "best_classifier": self._best_classifier,
+            "best_gap": self._best_gap,
+            "classifiers": self._classifiers,
+            "weights": self._weights,
+            "last_t": self._last_t,
+            "best_t": self._best_t,
+            "n_oracle_calls": self._n_oracle_calls
+        }
+
+
+class ExponentiatedGradient(ReductionsEstimator):
+    def __init__(self, estimator, disparity_metric, quality_metric, eps=0.01):
+        self._estimator = estimator
+        self._disparity_metric = disparity_metric
+        # TODO use quality metric
+        self._quality_metric = quality_metric
+        self._eps = eps
+        self._best_classifier = None
+        self._classifiers = None
+
+    def fit(self, X, y, aux_data=None, **kwargs):
+        # TODO: validate input data; unify between grid search an expgrad?
+        self._expgrad_result = exponentiated_gradient(X, aux_data, y, self._estimator,
+                                                      constraints=self._disparity_metric,
+                                                      eps=self._eps)
+        self._best_classifier = expgrad_result.best_classifier
+        self._classifiers = expgrad_result.classifiers
+        # TODO: figure out whether we should keep the remaining data of the result object
+
+    def predict(self, X):
+        return self._expgrad_result._predict_with_best_classifier(X)
```

Comment:
Shouldn't this be doing the stochastic one - the one which actually minimises disparity (on average)? Unless we have a requirement to make `predict` deterministic?

Reply:
@MiroDudik ?

Reply:
Right, so you say I should take this output number, take a random number and if the random number's above the output number then it's 1 else 0? (or I guess the other way round?)

Reply:
I thought it was that you picked one of the predictors in the result at random, and called predict on that.

Reply:
I may have confused best_classifier with that. It's just the mean_prediction of all classifiers. Actually, that should be equivalent since it's weighted. So that should be a float number, and I can interpret that as probability for 1, right?

Reply:
I think this is a @MiroDudik  question

Reply:
"predict" should be a randomized prediction, i.e., 0/1, whereas best_classifier is returning a real number between 0 and 1-corresponding to "posterior_predict". you should be flipping a coin based on what best_classifier is returning and returning the result of that flip as "predict".

Reply:
This should be correct now.